proposals/003-authentication-api.md

```diff
@@ -0,0 +1,223 @@
+            loginContext.login();
+            // here we propagate the subject along the pipeline
+            // using the new clientAuthentication() method which
+            // broadcasts the subject to all plugins in the upstream direction
+            context.clientAuthenticationSuccess(loginContext.getSubject());
+            return null; // TODO return a success response to the client
+        }
+        catch (LoginException e) {
+            // here we propagate the failure along the pipeline
+            // using the new clientAuthenticationFailure() method which
+            // broadcasts some represnetation of the error
+            // to all plugins in the upstream direction
+            context.clientAuthenticationFailure(e);
+            return null; // TODO return an error response to the client
+        }
+    }
+}
+```
+
+Notes:
+
+* by implementing `ClientSubjectAwareFilter` and reusing the `clientSubject`, it's possible to add a username principal to any existing `X500Principal` added to the object by the runtime.
+* `org.apache.kafka.common.security.plain.PlainLoginModule` (and probably other Kafka `LoginModule` implementations) don't work with how JASS `LoginModules` were originally designed to support stackable modules. For example it adds the `username` as a _credential_ of the `Subject`, rather than adding it as _principal_, and `login` is hard coded to return `true`.
+
+
+### Design choices
+
+* Use JAAS because:
+    - the Kafka broker and clients already use JAAS
+        - we'd rather avoid adding a dependency on those not-publicly-supportewd Kafka classes in kroxylicious-api and kroxylicious-runtime, but
+        - we recognise that 3rd part Filter authors might want to make that choice
+    - we have no appetite to build-out our own API, nor to pick up a dependency on a 3rd party framework
+* Use `Subject` to convey identity information, rather than `Principal` because 
+    - this is more sympathetic with the conceptpaul model of JAAS.
+    - it allows attaching multiple `Principals` to client `Subjects`, and to be consumed by `ClientSubjectAwareFilter` (e.g. know the client TLS certificate DN _and_ the SASL SCRAM-SHA username).
+
+## Affected/not affected projects
+
+The `kroxylicous` repo.
+
+## Compatibility
+
+This change would be backwards compatible for Filter developers and proxy users (i.e. all existing proxy configurations files would still be valid).
+
+## Rejected alternatives
+
+* Why not just add `public Principal clientPrincipal();` to `FilterContext`?
+    - It doesn't support multiple principals.
+* OK, so why not just add `public Subject clientSubject();` to `FilterContext`?
+    - It makes the `Subject` a property of the connection/Netty channel. The public API presented here allows a `Filter` to change the `Principals` presented to subsequent `Filters` in the chain, enabling Filters to implement use cases like impersonation.
```

Comment:
This assertion kinda makes me feel like we should be defining some standard or expected principals. Such as a KafkaPrincipal to represent the principal which has actually been authenticated by the broker.

Reply:
The latest update has specific Principal types defined. Marking resolved.